sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -3,16 +3,94 @@
 #
 #    Author: Julien Jerphanion <git@jjerphan.xyz>
 #
+# Overview
+# --------
 #
-# The abstractions defined here are used in various algorithms performing
-# the same structure of operations on distances between row vectors
-# of a datasets pair (X, Y).
+#    This module provides routines to compute pairwise distances between a set
+#    of row vectors of X and another set of row vectors of Y and apply a
+#    reduction on top. The canonical example is the brute-force computation
+#    of the top k nearest neighbors by leveraging the arg-k-min reduction.
 #
-# Importantly, the core of the computation is chunked to make sure that the pairwise
-# distance chunk matrices stay in CPU cache before applying the final reduction step.
-# Furthermore, the chunking strategy is also used to leverage OpenMP-based parallelism
-# (using Cython prange loops) which gives another multiplicative speed-up in
-# favorable cases on many-core machines.
+#    The reduction takes a matrix of pairwise distances between rows of X and Y
+#    as input and outputs an aggregate data-structure for each row of X. The
+#    aggregate values are typically smaller than the number of rows in Y, hence
+#    the term reduction.
+#
+#    For computational reasons, the reduction are performed on the fly on chunks
+#    of rows of X and Y so as to keep intermediate data-structures in CPU cache
+#    and avoid unnecessary round trips of large distance arrays with the RAM
+#    that would otherwise severely degrade the speed by making the overall
+#    processing memory-bound.
+#
+#    Finally, the routines follow a generic parallelization template to process
+#    chunks of data with OpenMP loops (via Cython prange), either on rows of X
+#    or rows of Y depending on their respective sizes.
+#
+#
+# Dispatching to specialized implementations
+# ------------------------------------------
+#
+#    Dispatchers are meant to be used in the Python code. Under the hood, a
+#    dispatcher must only define the logic to choose at runtime to the correct
+#    dtype-specialized :class:`PairwiseDistancesReduction` implementation based
+#    on the dtype of X and of Y.
+#
+#
+# High-level diagram
+# ------------------
+#
+#   Legend:
+#
+#      A ---⊳ B: A inherits from B
+#      A ---x B: A dispatches on B
+#      A ---* B: A composes B
+#
+#
+#                           (base dispatcher)
+#                       PairwiseDistancesReduction
+#                                   ∆
+#                                   |
+#                                   |
+#                 +-----------------+-----------------+
+#                 |                                   |
+#           (dispatcher)                        (dispatcher)
+#     PairwiseDistancesArgKmin           PairwiseDistancesRadiusNeighbors
```

Comment:
Are there future plans allow a third party to register a `ThirdPartyPairwiseDistancesArgKmin` with `PairwiseDistancesArgKmin` to dispatch to?

Reply:
That's an option, yet we need to discuss if upcoming third-party implementations will be higher level (mainly using numpy/scipy, etc.) or if they will be lower-level (mainly using Cython/C++ libraries, etc.). In that case, such dispatchers might be used (see the design note.

IMO this is just one option and the overall design needs more collective thoughts and discussions.

Reply:
Probably it now makes more sense to move this comment https://github.com/scikit-learn/scikit-learn/blob/c7dc9871874d6fb7d01563962914c4a8acc7b838/sklearn/metrics/_pairwise_distances_reduction.pyx#L1053-L1057

here.

Reply:
I agree with moving that part here.